File: api/internal/storage/repository/mocks.go
package repository

import (
	"github.com/georgi-georgiev/testmesh/internal/storage/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

// MockRepository handles mock server database operations
type MockRepository struct {
	db *gorm.DB
}

// NewMockRepository creates a new mock repository
func NewMockRepository(db *gorm.DB) *MockRepository {
	return &MockRepository{db: db}
}

// CreateServer creates a new mock server
func (r *MockRepository) CreateServer(server *models.MockServer) error {
	return r.db.Create(server).Error
}

// GetServerByID retrieves a mock server by ID
func (r *MockRepository) GetServerByID(id uuid.UUID) (*models.MockServer, error) {
	var server models.MockServer
	if err := r.db.First(&server, "id = ?", id).Error; err != nil {
		return nil, err
	}
	return &server, nil
}

// GetServerByPort retrieves a mock server by port
func (r *MockRepository) GetServerByPort(port int) (*models.MockServer, error) {
	var server models.MockServer
	if err := r.db.Where("port = ? AND status = ?", port, models.MockServerStatusRunning).First(&server).Error; err != nil {
		return nil, err
	}
	return &server, nil
}

// UpdateServer updates a mock server
func (r *MockRepository) UpdateServer(server *models.MockServer) error {
	return r.db.Save(server).Error
}

// ListServers retrieves mock servers with optional filters
func (r *MockRepository) ListServers(executionID *uuid.UUID, status models.MockServerStatus, limit, offset int) ([]models.MockServer, int64, error) {
	var servers []models.MockServer
	var total int64

	query := r.db.Model(&models.MockServer{})

	if executionID != nil {
		query = query.Where("execution_id = ?", *executionID)
	}
	if status != "" {
		query = query.Where("status = ?", status)
	}

	// Get total count
	if err := query.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	// Get paginated results
	if err := query.Order("created_at DESC").Limit(limit).Offset(offset).Find(&servers).Error; err != nil {
		return nil, 0, err
	}

	return servers, total, nil
}

// CreateEndpoint creates a new mock endpoint
func (r *MockRepository) CreateEndpoint(endpoint *models.MockEndpoint) error {
	return r.db.Create(endpoint).Error
}

// GetEndpointByID retrieves a mock endpoint by ID
func (r *MockRepository) GetEndpointByID(id uuid.UUID) (*models.MockEndpoint, error) {
	var endpoint models.MockEndpoint
	if err := r.db.First(&endpoint, "id = ?", id).Error; err != nil {
		return nil, err
	}
	return &endpoint, nil
}

// ListEndpoints retrieves endpoints for a mock server
func (r *MockRepository) ListEndpoints(serverID uuid.UUID) ([]models.MockEndpoint, error) {
	var endpoints []models.MockEndpoint
	if err := r.db.Where("mock_server_id = ?", serverID).Order("priority DESC, created_at ASC").Find(&endpoints).Error; err != nil {
		return nil, err
	}
	return endpoints, nil
}

// UpdateEndpoint updates a mock endpoint
func (r *MockRepository) UpdateEndpoint(endpoint *models.MockEndpoint) error {
	return r.db.Save(endpoint).Error
}

// DeleteEndpoint deletes a mock endpoint
func (r *MockRepository) DeleteEndpoint(id uuid.UUID) error {
	return r.db.Delete(&models.MockEndpoint{}, "id = ?", id).Error
}

// CreateRequest creates a new mock request log
func (r *MockRepository) CreateRequest(request *models.MockRequest) error {
	return r.db.Create(request).Error
}

// ListRequests retrieves requests for a mock server
func (r *MockRepository) ListRequests(serverID uuid.UUID, matched *bool, limit, offset int) ([]models.MockRequest, int64, error) {
	var requests []models.MockRequest
	var total int64

	query := r.db.Model(&models.MockRequest{}).Where("mock_server_id = ?", serverID)

	if matched != nil {
		query = query.Where("matched = ?", *matched)
	}

	// Get total count
	if err := query.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	// Get paginated results
	if err := query.Order("received_at DESC").Limit(limit).Offset(offset).Find(&requests).Error; err != nil {
		return nil, 0, err
	}

	return requests, total, nil
}

// GetState retrieves state for a mock server
func (r *MockRepository) GetState(serverID uuid.UUID, stateKey string) (*models.MockState, error) {
	var state models.MockState
	if err := r.db.Where("mock_server_id = ? AND state_key = ?", serverID, stateKey).First(&state).Error; err != nil {
		return nil, err
	}
	return &state, nil
}

// UpsertState creates or updates state for a mock server
func (r *MockRepository) UpsertState(state *models.MockState) error {
	return r.db.Save(state).Error
}

// ListStates retrieves all states for a mock server
func (r *MockRepository) ListStates(serverID uuid.UUID) ([]models.MockState, error) {
	var states []models.MockState
	if err := r.db.Where("mock_server_id = ?", serverID).Find(&states).Error; err != nil {
		return nil, err
	}
	return states, nil
}

// DeleteServer deletes a mock server (cascades to endpoints, requests, state)
func (r *MockRepository) DeleteServer(id uuid.UUID) error {
	return r.db.Delete(&models.MockServer{}, "id = ?", id).Error
}

// DeleteState deletes a state entry for a mock server
func (r *MockRepository) DeleteState(serverID uuid.UUID, stateKey string) error {
	return r.db.Delete(&models.MockState{}, "mock_server_id = ? AND state_key = ?", serverID, stateKey).Error
}

// GetRequestByID retrieves a specific mock request by ID
func (r *MockRepository) GetRequestByID(id uuid.UUID) (*models.MockRequest, error) {
	var request models.MockRequest
	if err := r.db.First(&request, "id = ?", id).Error; err != nil {
		return nil, err
	}
	return &request, nil
}

// DeleteRequests deletes all request logs for a mock server
func (r *MockRepository) DeleteRequests(serverID uuid.UUID) error {
	return r.db.Delete(&models.MockRequest{}, "mock_server_id = ?", serverID).Error
}

// CreateState creates a new state entry for a mock server
func (r *MockRepository) CreateState(state *models.MockState) error {
	return r.db.Create(state).Error
}

File: dashboard/lib/api/mocks.ts
import axios from 'axios';
import type {
  MockServer,
  MockEndpoint,
  MockRequest,
  MockState,
  ListMockServersResponse,
  ListMockEndpointsResponse,
  ListMockRequestsResponse,
  ListMockStatesResponse,
  MockServerStatus,
} from './types';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:5016';

const apiClient = axios.create({
  baseURL: API_BASE_URL,
  timeout: 30000,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Mock Server API
export const mockServerApi = {
  // List mock servers
  list: async (params?: {
    execution_id?: string;
    status?: MockServerStatus;
    limit?: number;
    offset?: number;
  }): Promise<ListMockServersResponse> => {
    const response = await apiClient.get<ListMockServersResponse>('/api/v1/mock-servers', {
      params,
    });
    return response.data;
  },

  // Get mock server by ID
  get: async (id: string): Promise<MockServer> => {
    const response = await apiClient.get<MockServer>(`/api/v1/mock-servers/${id}`);
    return response.data;
  },

  // Create a new standalone mock server
  create: async (data: { name: string }): Promise<MockServer> => {
    const response = await apiClient.post<MockServer>('/api/v1/mock-servers', data);
    return response.data;
  },

  // Delete mock server
  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/v1/mock-servers/${id}`);
  },

  // Start a stopped mock server
  start: async (id: string): Promise<MockServer> => {
    const response = await apiClient.post<MockServer>(`/api/v1/mock-servers/${id}/start`);
    return response.data;
  },

  // Stop a running mock server
  stop: async (id: string): Promise<MockServer> => {
    const response = await apiClient.post<MockServer>(`/api/v1/mock-servers/${id}/stop`);
    return response.data;
  },

  // Get endpoints for a mock server
  getEndpoints: async (serverId: string): Promise<ListMockEndpointsResponse> => {
    const response = await apiClient.get<ListMockEndpointsResponse>(
      `/api/v1/mock-servers/${serverId}/endpoints`
    );
    return response.data;
  },

  // Create endpoint
  createEndpoint: async (serverId: string, endpoint: Partial<MockEndpoint>): Promise<MockEndpoint> => {
    const response = await apiClient.post<MockEndpoint>(
      `/api/v1/mock-servers/${serverId}/endpoints`,
      endpoint
    );
    return response.data;
  },

  // Update endpoint
  updateEndpoint: async (
    serverId: string,
    endpointId: string,
    endpoint: Partial<MockEndpoint>
  ): Promise<MockEndpoint> => {
    const response = await apiClient.put<MockEndpoint>(
      `/api/v1/mock-servers/${serverId}/endpoints/${endpointId}`,
      endpoint
    );
    return response.data;
  },

  // Delete endpoint
  deleteEndpoint: async (serverId: string, endpointId: string): Promise<void> => {
    await apiClient.delete(`/api/v1/mock-servers/${serverId}/endpoints/${endpointId}`);
  },

  // Get request logs
  getRequests: async (
    serverId: string,
    params?: {
      matched?: boolean;
      limit?: number;
      offset?: number;
    }
  ): Promise<ListMockRequestsResponse> => {
    const response = await apiClient.get<ListMockRequestsResponse>(
      `/api/v1/mock-servers/${serverId}/requests`,
      { params }
    );
    return response.data;
  },

  // Get server state
  getStates: async (serverId: string): Promise<ListMockStatesResponse> => {
    const response = await apiClient.get<ListMockStatesResponse>(
      `/api/v1/mock-servers/${serverId}/state`
    );
    return response.data;
  },

  // Get specific state by key
  getState: async (serverId: string, key: string): Promise<MockState> => {
    const response = await apiClient.get<MockState>(
      `/api/v1/mock-servers/${serverId}/state/${key}`
    );
    return response.data;
  },

  // Create state
  createState: async (
    serverId: string,
    data: { state_key: string; state_value: any }
  ): Promise<MockState> => {
    const response = await apiClient.post<MockState>(
      `/api/v1/mock-servers/${serverId}/state`,
      data
    );
    return response.data;
  },

  // Update state
  updateState: async (
    serverId: string,
    key: string,
    data: { state_value: any }
  ): Promise<MockState> => {
    const response = await apiClient.put<MockState>(
      `/api/v1/mock-servers/${serverId}/state/${key}`,
      data
    );
    return response.data;
  },

  // Delete state
  deleteState: async (serverId: string, key: string): Promise<void> => {
    await apiClient.delete(`/api/v1/mock-servers/${serverId}/state/${key}`);
  },

  // Get specific request by ID
  getRequest: async (serverId: string, requestId: string): Promise<MockRequest> => {
    const response = await apiClient.get<MockRequest>(
      `/api/v1/mock-servers/${serverId}/requests/${requestId}`
    );
    return response.data;
  },

  // Clear all request logs
  clearRequests: async (serverId: string): Promise<void> => {
    await apiClient.delete(`/api/v1/mock-servers/${serverId}/requests`);
  },
};

export default mockServerApi;
